File: common/src/main/java/juuxel/adorn/client/gui/screen/MainConfigScreen.java
package juuxel.adorn.client.gui.screen;

import juuxel.adorn.client.gui.widget.Panel;
import juuxel.adorn.config.ConfigManager;
import juuxel.adorn.fluid.FluidUnit;
import juuxel.adorn.item.group.ItemGroupingOption;
import juuxel.adorn.util.PropertyRef;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.gui.components.Button;
import net.minecraft.network.chat.CommonComponents;
import net.minecraft.network.chat.Component;

import java.util.Arrays;

public final class MainConfigScreen extends AbstractConfigScreen {
    private static final int BUTTON_WIDTH = 200;
    private static final Layout LAYOUT = new Layout.BigButtons(BUTTON_WIDTH);

    public MainConfigScreen(Screen parent) {
        super(Component.translatable("gui.adorn.config.title"), parent, LAYOUT);
    }

    @Override
    protected void init() {
        super.init();
        addRenderableWidget(
            Button.builder(CommonComponents.GUI_DONE, widget -> onClose())
                .pos(width / 2 - 100, height - BACK_BUTTON_Y_FROM_BOTTOM)
                .size(200, 20)
                .build()
        );
    }

    @Override
    protected void initConfigWidgets(Panel panel) {
        var config = ConfigManager.config();
        int x = (width - BUTTON_WIDTH) / 2;
        addHeading(Component.translatable("gui.adorn.config.visual"));
        addConfigToggle(PropertyRef.ofField(config.client, "showTradingStationTooltips"));
        addConfigButton(PropertyRef.ofField(config.client, "displayedFluidUnit"), Arrays.asList(FluidUnit.values()));
        addHeading(Component.translatable("gui.adorn.config.creative_inventory"));
        addConfigToggle(PropertyRef.ofField(config.client, "showItemsInStandardGroups"));
        addConfigButton(
            PropertyRef.ofField(config, "groupItems"),
            Arrays.asList(ItemGroupingOption.values()),
            true
        );
        addHeading(Component.translatable("gui.adorn.config.other"));
        addSubscreenButton(Component.translatable("gui.adorn.config.game_rule_defaults"), GameRuleDefaultsScreen::new);
        addSubscreenButton(Component.translatable("gui.adorn.config.toggle_mod_compatibility"), ModCompatConfigScreen::new);
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/screen/MainConfigScreen.kt
package juuxel.adorn.client.gui.screen

import juuxel.adorn.client.gui.widget.ConfigScreenHeading
import juuxel.adorn.config.ConfigManager
import juuxel.adorn.fluid.FluidUnit
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.ButtonWidget
import net.minecraft.screen.ScreenTexts
import net.minecraft.text.Text

class MainConfigScreen(parent: Screen) : AbstractConfigScreen(Text.translatable("gui.adorn.config.title"), parent) {
    override fun init() {
        super.init()
        val config = ConfigManager.config()
        val x = (width - BUTTON_WIDTH) / 2
        var nextY = 40
        addDrawable(ConfigScreenHeading(Text.translatable("gui.adorn.config.visual"), x, nextY, BUTTON_WIDTH))
        nextY += ConfigScreenHeading.HEIGHT
        addDrawableChild(createConfigToggle(x, nextY, BUTTON_WIDTH, config.client::showTradingStationTooltips))
        nextY += BUTTON_SPACING
        addDrawableChild(createConfigToggle(x, nextY, BUTTON_WIDTH, config.client::showItemsInStandardGroups))
        nextY += BUTTON_SPACING
        addDrawableChild(createConfigButton(x, nextY, BUTTON_WIDTH, config.client::displayedFluidUnit, FluidUnit.values().toList()))
        nextY += BUTTON_SPACING
        addDrawable(ConfigScreenHeading(Text.translatable("gui.adorn.config.other"), x, nextY, BUTTON_WIDTH))
        nextY += ConfigScreenHeading.HEIGHT
        addDrawableChild(
            ButtonWidget(x, nextY, BUTTON_WIDTH, 20, Text.translatable("gui.adorn.config.game_rule_defaults")) {
                client!!.setScreen(GameRuleDefaultsScreen(this))
            }
        )
        addDrawableChild(ButtonWidget(this.width / 2 - 100, this.height - 27, 200, 20, ScreenTexts.DONE) { close() })
    }

    companion object {
        private const val HEADING_HEIGHT = 12
        private const val BUTTON_WIDTH = 200
    }
}
